src/main/scala/objsets/TweetSet.scala

```diff
@@ -0,0 +1,254 @@
+    *
+    * Question: Should we implement this method here, or should it remain abstract
+    * and be implemented in the subclasses?
+    */
+  def union(that: TweetSet): TweetSet
+
+  /**
+    * Returns the tweet from this set which has the greatest retweet count.
+    *
+    * Calling `mostRetweeted` on an empty set should throw an exception of
+    * type `java.util.NoSuchElementException`.
+    *
+    * Question: Should we implement this method here, or should it remain abstract
+    * and be implemented in the subclasses?
+    */
+  def mostRetweeted: Tweet
+
+  /**
+    * Returns a list containing all tweets of this set, sorted by retweet count
+    * in descending order. In other words, the head of the resulting list should
+    * have the highest retweet count.
+    *
+    * Hint: the method `remove` on TweetSet will be very useful.
+    * Question: Should we implement this method here, or should it remain abstract
+    * and be implemented in the subclasses?
+    */
+  def descendingByRetweet: TweetList =
+    if (isEmpty) {
+      Nil
+    } else {
+      val twt = mostRetweeted
+      new Cons(twt, remove(twt).descendingByRetweet)
+    }
+
+  def isEmpty: Boolean
+
+  /**
+    * The following methods are already implemented
+    */
+  /**
+    * Returns a new `TweetSet` which contains all elements of this set, and the
+    * the new element `tweet` in case it does not already exist in this set.
+    *
+    * If `this.contains(tweet)`, the current set is returned.
+    */
+  def incl(tweet: Tweet): TweetSet
+
+  /**
+    * Returns a new `TweetSet` which excludes `tweet`.
+    */
+  def remove(tweet: Tweet): TweetSet
+
+  /**
+    * Tests if `tweet` exists in this `TweetSet`.
+    */
+  def contains(tweet: Tweet): Boolean
+
+  /**
+    * This method takes a function and applies it to every element in the set.
+    */
+  def foreach(f: Tweet => Unit): Unit
+}
+
+class Empty extends TweetSet {
+
+  def filterAcc(p: Tweet => Boolean, acc: TweetSet): TweetSet = acc
+
+  def union(that: TweetSet): TweetSet = that
+
+  def mostRetweeted: Tweet = throw new NoSuchElementException
+
+  def isEmpty: Boolean = true
+
+  /**
+    * The following methods are already implemented
+    */
+  def contains(tweet: Tweet): Boolean = false
+
+  def incl(tweet: Tweet): TweetSet = new NonEmpty(tweet, new Empty, new Empty)
+
+  def remove(tweet: Tweet): TweetSet = this
+
+  def foreach(f: Tweet => Unit): Unit = ()
+}
+
+class NonEmpty(elem: Tweet, left: TweetSet, right: TweetSet) extends TweetSet {
+
+  def filterAcc(p: Tweet => Boolean, acc: TweetSet): TweetSet =
+    if (p(elem)) {
+      left.filterAcc(p, right.filterAcc(p, acc.incl(elem)))
+    } else {
+      left.filterAcc(p, right.filterAcc(p, acc))
+    }
+
+  def union(that: TweetSet): TweetSet =
+    (left.union(right.union(that))).incl(elem)
+
+  def isEmpty: Boolean = false
+
+  def mostRetweeted: Tweet =
+    if (!left.isEmpty && left.mostRetweeted.retweets > elem.retweets) {
```

Comment:
You can use `left.nonEmpty`.

Reply:
The class `TweetSet` does not contain the method called `nonEmpty`.